loopd/daemon.go

```diff
@@ -161,75 +184,191 @@ func daemon(config *config, lisCfg *listenerCfg) error {
+		}
+	}
 
-		log.Infof("RPC server listening on %s", grpcListener.Addr())
+	log.Infof("Swap server address: %v", d.cfg.SwapServer)
 
-		if restListener != nil {
-			log.Infof("REST proxy listening on %s", restListener.Addr())
+	// Create an instance of the loop client library.
+	swapclient, clientCleanup, err := getClient(d.cfg, &d.lnd.LndServices)
+	if err != nil {
+		return err
+	}
+
+	// Now finally fully initialize the swap client RPC server instance.
+	d.swapClientServer = swapClientServer{
+		impl:        swapclient,
+		lnd:         &d.lnd.LndServices,
+		swaps:       make(map[lntypes.Hash]loop.SwapInfo),
+		subscribers: make(map[int]chan<- interface{}),
+		statusChan:  make(chan loop.SwapInfo),
+	}
+	d.clientCleanup = clientCleanup
+
+	// Retrieve all currently existing swaps from the database.
+	swapsList, err := d.impl.FetchSwaps()
+	if err != nil {
+		return err
+	}
+
+	for _, s := range swapsList {
+		d.swaps[s.SwapHash] = *s
+	}
+
+	// Start the main event handlers, now that we have prepared everything
+	// they need to work.
+	d.startSwapHandlers()
+
+	return nil
+}
+
+// startSwapHandlers starts all goroutines that are tasked with handling swap
+// events.
+func (d *Daemon) startSwapHandlers() {
+	d.mainCtx, d.mainCtxCancel = context.WithCancel(context.Background())
+
+	// Start the swap client itself.
+	d.wg.Add(1)
+	go func() {
+		defer d.wg.Done()
+
+		log.Infof("Starting swap client")
+		err := d.impl.Run(d.mainCtx, d.statusChan)
+		if err != nil {
+			log.Error(err)
 		}
+		log.Infof("Swap client stopped")
 
-		err = grpcServer.Serve(grpcListener)
+		// In case we got here because the swap client crashed, we need
+		// to shutdown as cleanly as possible. If we got here because of
+		// a normal shutdown, stop will prevent itself from running
+		// twice.
+		err = d.Stop()
 		if err != nil {
 			log.Error(err)
 		}
 	}()
 
-	interruptChannel := make(chan os.Signal, 1)
-	signal.Notify(interruptChannel, os.Interrupt)
+	// Start a goroutine that broadcasts swap updates to clients.
+	d.wg.Add(1)
+	go func() {
+		defer d.wg.Done()
+
+		log.Infof("Waiting for updates")
+		d.processStatusUpdates(d.mainCtx)
+	}()
+}
+
+// Stop tries to gracefully shut down the daemon.
+func (d *Daemon) Stop() error {
+	// If the swap client crashes, stop is called to make sure everything
+	// is cleaned up. But this also means we get here twice in case of the
+	// normal shutdown because after the swap client exists normally, stop
+	// is called again. We don't want to log an error in that case.
+	if atomic.AddInt32(&d.stopped, 1) != 1 {
+		return nil
+	}
 
-	// Run until the users terminates loopd or an error occurred.
-	select {
-	case <-interruptChannel:
-		log.Infof("Received SIGINT (Ctrl+C).")
+	// First of all, cancel the main context that all event handlers are
+	// using. This should stop all swap activity and all event handlers
+	// should exit.
+	if d.mainCtxCancel != nil {
+		d.mainCtxCancel()
+	}
```

Comment:
Can you just do this, stop everything at the same time? I would think that services need to be stopped outside to inside and wait for each one to have stopped properly.

In general I feel that the `Start`/`Stop` pattern makes the whole shutdown sequence less readable and forces us to resort to these nil checks, where as otherwise we could just keep adding defers.

How about keeping a synchronous Run() method that does all its clean up and adding a wrapping `Start`/`Stop` construction that just initiates a goroutine?

Reply:
Discussed this offline. Decided to stay with the `Start`/`Stop` model but add an error channel that can be used to detect runtime errors.

I did some manual tests and made sure that the daemon stops correctly if the swap handler crashes and also if a `monitor` subscription is connected.
For the second case I also fixed it to return a nice error message instead of just force closing the gRPC server and returning a "connection is terminating" error.